core/runtime.rs

```diff
@@ -1566,7 +1564,7 @@ impl JsRuntimeForSnapshot {
     // Serialize the module map and store its data in the snapshot.
     {
       let snapshotted_data = {
-        let module_map_rc = self.module_map.take().unwrap();
+        let module_map_rc = std::mem::take(&mut self.module_map);
```

Comment:
Not totally sure, but since it's an Rc<...> could this just be self.module_map.clone()? Otherwise we're constructing a default just to drop it shortly after.

Reply:
Hmm I thought this code was to do with cleaning up for snapshotting but that's done elsewhere...

I think the `Option` was needed to make it clear that this slot no longer corresponds with the context slot at `MODULE_MAP_DATA_OFFSET`, so removing the option probably isn't a great change.

Reply:
I think it'll be OK to just use a ref or clone here because it's going to be alive until prepare_for_snapshot below, and this method does not appear to consume the module map. Everything in the PR looks good but I think we can optimize creating this default when we don't have to.

Reply:
Yeah I realised the previous value of the rc would get dropped anyway at the end of this block so it doesn't stop the module map from dropping twice.

@bartlomieju Do you know what the purpose was of the `.take()` here?

Reply:
Oh @mmastrac we do it to ensure the old values (which are in isolate slots) are dropped before the isolate is destroyed as I said, but the reason it's not erroring is because we deliberately leak the isolate when snapshotting. Not sure what we should do in that case.

Reply:
This is actually fine as we call `cleanup` as part of `prepare_for_snapshot` (see the last statement of `snapshot()`) which drops the old value.

Reply:
So basically the isolate is cleaned out (including the slots), all of the `Rc`s are dropped, then that isolate is "leaked" to the`SnapshotCreator`. 

We _should_ probably assert that the `Rc` count of module map is 1 in the `cleanup` function.

Reply:
> This is actually fine as we call `cleanup` as part of `prepare_for_snapshot` (see the last statement of `snapshot()`) which drops the slot-stored value.

I'm aware that `cleanup()` restores and then drops an rc pointing at the module map, but currently in this PR this is happening while the other rc is still held so it does nothing.

Reply:
> We _should_ probably assert that the `Rc` count of module map is 1 in the `cleanup` function.

Yep I tried it locally this assertion fails

Reply:
Ah ok. Sounds good -- let me know when you're ready to merge. Looking great and thanks for the additional cleanup.

Reply:
Thanks, it's ready